pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/ReplicaGroupRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,573 @@
+    List<String> serverInstances = replicaGroupPartitionAssignment.getAllInstances();
+
+    List<String> addedServers = new ArrayList<>();
+    List<String> removedServers = new ArrayList<>();
+    for (String server : oldServerInstances) {
+      if (!serverInstances.contains(server)) {
+        removedServers.add(server);
+      }
+    }
+    for (String server : serverInstances) {
+      if (!oldServerInstances.contains(server)) {
+        addedServers.add(server);
+      }
+    }
+
+    // Add servers to the mapping
+    for (String server : addedServers) {
+      serverToSegments.put(server, new LinkedList<String>());
+    }
+
+    // Remove servers from the mapping
+    for (String server : removedServers) {
+      serverToSegments.remove(server);
+    }
+
+    // Check if rebalance can cause the downtime
+    Set<String> segmentsToCover = segmentToServerMapping.keySet();
+    Set<String> coveredSegments = new HashSet<>();
+    for (Map.Entry<String, LinkedList<String>> entry : serverToSegments.entrySet()) {
+      coveredSegments.addAll(entry.getValue());
+    }
+    if (!segmentsToCover.equals(coveredSegments)) {
+      LOGGER.warn("Segment is lost, the rebalance will cause downtime of the segment.");
+    }
+
+    ReplicaGroupStrategyConfig replicaGroupConfig = tableConfig.getValidationConfig().getReplicaGroupStrategyConfig();
+    // If no replica group config is available, we cannot perform the rebalance algorithm
+    if (replicaGroupConfig == null || replicaGroupConfig.getPartitionColumn() != null) {
+      throw new UnsupportedOperationException(
+          "This table is not using replica group segment assignment or table level replica group");
+    }
+
+    boolean mirrorAssignment = replicaGroupConfig.getMirrorAssignmentAcrossReplicaGroups();
+    int numPartitions = replicaGroupPartitionAssignment.getNumPartitions();
+    int numReplicaGroups = replicaGroupPartitionAssignment.getNumReplicaGroups();
+
+    for (int partitionId = 0; partitionId < numPartitions; partitionId++) {
+      List<String> referenceReplicaGroup = new ArrayList<>();
+      for (int replicaId = 0; replicaId < numReplicaGroups; replicaId++) {
+        List<String> serversInReplicaGroup =
+            replicaGroupPartitionAssignment.getInstancesfromReplicaGroup(partitionId, replicaId);
+        if (replicaId == 0) {
+          // We need to keep the first replica group in case of mirroring.
+          referenceReplicaGroup.addAll(serversInReplicaGroup);
+        } else if (mirrorAssignment) {
+          // Copy the segment assignment from the reference replica group
+          for (int i = 0; i < serversInReplicaGroup.size(); i++) {
+            serverToSegments.put(serversInReplicaGroup.get(i), serverToSegments.get(referenceReplicaGroup.get(i)));
+          }
+          continue;
+        }
+
+        // Make sure that all the segment is covered only once within a replica group.
```

Comment:
clarification question so I can wrap my head around this. Unless we are "bootstrapping", I suppose all segments will be covered exactly once amongst the servers?

Reply:
I recently added the code to update replica group mapping(INSTANCE_PARTITION) when the user updates the table config from non-replica group to replica group aware. Imagine that some segments get pushed before you run the rebalancer. Then, all the config will indicate that you have the correct replica group configs while some segments are assigned based on default strategy and others are assigned based on replica group strategy. In this case, the rebalancer will behave in a weird way if the user do not use "bootstrap" mode explicitly. Solution to this problem will be either I make sure all the segment is covered only for once within the code, or I can throw the exception and provide the message to use "bootstrap" mode.